python/lsst/ip/diffim/zogy.py

```diff
@@ -330,748 +127,715 @@ def _computeVarianceMean(self, exposure):
+                PSF of ``D``.
+            - ``S`` : `numpy.ndarray` of `numpy.complex` or `None`
+                Significance (score) image.
+            - ``varplaneS`` : `numpy.ndarray` of `numpy.complex` or `None`
+                Variance plane of ``S``.
+            - ``Ps`` : `numpy.ndarray` of `numpy.complex`
+                PSF of ``S``.
 
-    def run(self, subExposure, expandedSubExposure, fullBBox, template,
-            **kwargs):
-        """Perform ZOGY proper image subtraction on sub-images
+        Notes
+        -----
+        All array inputs and outputs are Fourier-space images with size of
+        `self.freqSpaceShape` in this method.
 
-        This method performs ZOGY proper image subtraction on
-        `subExposure` using local measures for image variances and
-        PSF. `subExposure` is a sub-exposure of the science image. It
-        also requires the corresponding sub-exposures of the template
-        (`template`). The operations are actually performed on
-        `expandedSubExposure` to allow for invalid edge pixels arising
-        from convolutions, which are then removed.
+        ``varMean1``, ``varMean2`` quantities are part of the noise model and not to be confused
+        with the variance of image frequency components or with ``varPlane1``, ``varPlane2`` that
+        are the Fourier transform of the variance planes.
+        """
+        var1F2Sq = varMean1*F2*F2
+        var2F1Sq = varMean2*F1*F1
+        # We need reals for comparison, also real operations are usually faster
+        psfAbsSq1 = np.real(np.conj(psf1)*psf1)
+        psfAbsSq2 = np.real(np.conj(psf2)*psf2)
+        FdDenom = np.sqrt(var1F2Sq + var2F1Sq)  # one number
+
+        # Secure positive limit to avoid floating point operations resulting in exact zero
+        tiny = np.finfo(psf1.dtype).tiny * 100
+        sDenom = var1F2Sq*psfAbsSq2 + var2F1Sq*psfAbsSq1  # array, eq. (12)
+        # Frequencies where both psfs are too close to zero.
+        # We expect this only in cases when psf1, psf2 are identical,
+        # and either having very well sampled Gaussian tails
+        # or having "edges" such that some sinc-like zero crossings are found at symmetry points
+        #
+        # if sDenom < tiny then it can be == 0. -> `denom` = 0. and 0/0 occur at `c1` , `c2`
+        # if we keep SDenom = tiny, denom ~ O(sqrt(tiny)), Pd ~ O(sqrt(tiny)), S ~ O(sqrt(tiny)*tiny) == 0
+        # Where S = 0 then Pd = 0 and D should still yield the same variance ~ O(1)
+        # For safety, we set S = 0 explicitly, too, though it should be unnecessary.
+        fltZero = sDenom < tiny
+        nZero = np.sum(fltZero)
+        self.log.debug(f"Handling {nZero} both PSFs are zero points.")
+        if nZero > 0:
+            fltZero = np.nonzero(fltZero)  # We expect only a small fraction of such frequencies
+            sDenom[fltZero] = tiny  # Avoid division problem but overwrite result anyway
```

Comment:
Yes, it works without converting the _bool_ into an _index array_ (of ints). This supposed to be a speed optimization. If only a small fraction of a large array is reassigned, it is faster to index the array with an array of ints rather than with a bool array.